### PR TITLE
many: allow ignoring running apps for specific request

### PR DESCRIPTION
**Pull request by zyga**
This patch adds support for "snap {install,refresh,revert}
--ignore-running" across API boundary, so that ultimately the two checks
for a snap being in use are not performed, even if the feature is
enabled.

This is primarily useful for spread tests, which assume the old behavior
and want to observe the in certain edge cases. It's not hidden as it may
be used by end users as well.

I left out snap {try,amend} even though they should have this option
as well, as none of the test required it. I'm looking for early feedback
on naming and implementation.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -833,7 +833,7 @@ func (m *SnapManager) doUnlinkCurrentSnap(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
-	if experimentalRefreshAppAwareness {
+	if experimentalRefreshAppAwareness && !snapsup.Flags.IgnoreRunning {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go:836]: this needs tests

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -174,7 +174,7 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		}
 		snapsup.PlugsOnly = snapsup.PlugsOnly && (len(info.Slots) == 0)
 
-		if experimentalRefreshAppAwareness {
+		if experimentalRefreshAppAwareness && !snapsup.Flags.IgnoreRunning {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:177]: same

File: daemon/api.go [hunk 3]
```diff
@@ -831,6 +832,10 @@ func (inst *snapInstruction) installFlags() (snapstate.Flags, error) {
 	if inst.Unaliased {
 		flags.Unaliased = true
 	}
+	if inst.IgnoreRunning {
+		flags.IgnoreRunning = true
+	}
```
Comment (pedronis) [inline daemon/api.go:837]: same

File: daemon/api.go [hunk 4]
```diff
@@ -1025,6 +1030,9 @@ func snapUpdate(inst *snapInstruction, st *state.State) (string, []*state.TaskSe
 	if inst.IgnoreValidation {
 		flags.IgnoreValidation = true
 	}
+	if inst.IgnoreRunning {
+		flags.IgnoreRunning = true
```
Comment (pedronis) [inline daemon/api.go:1034]: same

File: cmd/snap/cmd_snap_op.go [hunk 5]
```diff
@@ -970,8 +977,9 @@ type cmdRevert struct {
 	waitMixin
 
 	modeMixin
-	Revision   string `long:"revision"`
-	Positional struct {
+	Revision      string `long:"revision"`
+	IgnoreRunning bool   `long:"ignore-running"`
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:981]: I think this should start hidden.  The name is reasonable but we will also add a way for snap to have a hook that can block refreshes, that might need  yet another flag,  I'm a bit unsure ATM whether we want then two flags or a single "force" flags that covers more.

**Review by pedronis — COMMENTED** [review]
thanks for this, did a pass, my main input is that the flag should start hidden

**Comment by zyga** [comment]
@pedronis thank you for the review. I'll make the flag hidden and iterate!

File: cmd/snap/cmd_snap_op.go (hunk 5 above)
Comment (zyga) [reply cmd/snap/cmd_snap_op.go:981]: I agree with the rationale, done.

File: daemon/api.go (hunk 3 above)
Comment (zyga) [reply daemon/api.go:837]: Done

File: daemon/api.go (hunk 4 above)
Comment (zyga) [reply daemon/api.go:1034]: Done

File: overlord/snapstate/handlers.go (hunk 1 above)
Comment (zyga) [reply overlord/snapstate/handlers.go:836]: Done

File: cmd/snap/cmd_snap_op.go [hunk 6]
```diff
@@ -637,6 +639,7 @@ type cmdRefresh struct {
 	List             bool   `long:"list"`
 	Time             bool   `long:"time"`
 	IgnoreValidation bool   `long:"ignore-validation"`
+	IgnoreRunning    bool   `long:"ignore-running"`
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go:642]: this needs hidden too

**Review by pedronis — COMMENTED** [review]
looks ok to me but missing hidden flag

File: cmd/snap/cmd_snap_op.go (hunk 6 above)
Comment (zyga) [reply cmd/snap/cmd_snap_op.go]: Done!

**Review by pedronis — APPROVED** [review]
thanks

File: client/snap_op.go [hunk 7]
```diff
@@ -74,6 +75,9 @@ func (opts *SnapOptions) writeModeFields(mw *multipart.Writer) error {
 }
 
 func (opts *SnapOptions) writeOptionFields(mw *multipart.Writer) error {
+	if err := writeFieldBool(mw, "ignore-running", opts.IgnoreRunning); err != nil {
```
Comment (zyga) [inline client/snap_op.go:78]: This is not tested

Comment (zyga) [reply client/snap_op.go:78]: Done

**Review by mvo5 — APPROVED** [review]
Thank you!

**Pull request merged**